src/IIIFPresentation/API/Features/Storage/StorageController.cs

```diff
@@ -20,39 +21,62 @@
 using Microsoft.AspNetCore.WebUtilities;
 using Microsoft.Extensions.Options;
 using Models.API.Collection;
-using Newtonsoft.Json;
+using Models.Database.General;
+using Repository;
+using Repository.Helpers;
 
 namespace API.Features.Storage;
 
 [Route("/{customerId:int}")]
 [ApiController]
-public class StorageController(IAuthenticator authenticator, IOptions<ApiSettings> options, IMediator mediator)
+public class StorageController(
+    IAuthenticator authenticator,
+    PresentationContext dbContext,
+    IOptions<ApiSettings> options,
+    IMediator mediator)
     : PresentationController(options.Value, mediator)
 {
-    private JsonSerializerSettings jsonSettings;
-    
     [HttpGet("{*slug}")]
     [ETagCaching]
     [VaryHeader]
-    public async Task<IActionResult> GetHierarchicalCollection(int customerId, string slug = "")
+    public async Task<IActionResult> GetHierarchical(int customerId, string slug = "")
     {
-        var storageRoot = await Mediator.Send(new GetHierarchicalCollection(customerId, slug, GetUrlRoots()));
+        var hierarchy =
+            await dbContext.RetrieveHierarchy(customerId, slug);
 
-        if (storageRoot.Collection is not { IsPublic: true }) return this.PresentationNotFound();
-
-        if (Request.HasShowExtraHeader() && await authenticator.ValidateRequest(Request) == AuthResult.Success)
+        switch (hierarchy?.Type)
         {
-            var relativeUrl = storageRoot.Collection.GenerateFlatCollectionId(GetUrlRoots());
-            relativeUrl = QueryHelpers.AddQueryString(relativeUrl, Request.Query);
-            return SeeOther(relativeUrl);
+            case ResourceType.IIIFManifest:
+                if (Request.HasShowExtraHeader())
+                    return hierarchy.ManifestId == null ? NotFound() : SeeOther($"manifests/{hierarchy.ManifestId}");
+
+                var storedManifest = await mediator.Send(new GetManifestHierarchical(hierarchy, slug, GetUrlRoots()));
+                return storedManifest == null ? NotFound() : Ok(storedManifest);
+
+            case ResourceType.IIIFCollection:
```

Comment:
Not a problem, just a comment - At first I thought this was an unused case, but I wasn't aware that multiple `case` statements could be used together - neat!